Mathlib/FieldTheory/Finite/Basic.lean

```diff
@@ -309,11 +312,111 @@ theorem sum_pow_lt_card_sub_one (i : ℕ) (h : i < q - 1) : ∑ x : K, x ^ i = 0
       _ = ∑ x : Kˣ, (x ^ i : K) := by simp [φ, ← this, univ.sum_map φ]
       _ = 0 := by rw [sum_pow_units K i, if_neg]; exact hiq
 
+section frobenius
+
+variable (R) [CommRing R] [Algebra K R]
+
+open Classical (indefiniteDescription) in
+private noncomputable def frobeniusAlgHomAux [Nontrivial R] :
+    {f : R →ₐ[K] R // ∀ r, f r = r ^ q} :=
+  let p := indefiniteDescription _ (card' K)
+  let n := indefiniteDescription _ p.2.2
+  have := p.2.1
+  have : ExpChar K p := ExpChar.prime n.2.1
+  have := expChar_of_injective_algebraMap (algebraMap K R).injective p
+  ⟨⟨iterateFrobenius R p n,
+    fun k ↦ (iterateFrobenius_def ..).trans <| by rw [← map_pow, ← n.2.2, pow_card]⟩,
+    fun r ↦ (iterateFrobenius_def ..).trans <| by rw [← n.2.2]⟩
+
+/-- If R is an algebra over a finite field K, the Frobenius K-algebra endomorphism of R is
+  given by raising every element of R to its #K-th power. -/
+@[simps!] def frobeniusAlgHom : R →ₐ[K] R where
+  __ := powMonoidHom q
+  map_zero' := zero_pow Fintype.card_pos.ne'
+  map_add' _ _ := show _ ^ _ = _ ^ _ + _ ^ _ by
+    nontriviality R
+    simp_rw [← (frobeniusAlgHomAux K R).2]
+    apply map_add
+  commutes' _ := show _ ^ _ = _ by
+    nontriviality R
+    rw [← (frobeniusAlgHomAux K R).2]
+    apply AlgHom.commutes
+
+theorem coe_frobeniusAlgHom : ⇑(frobeniusAlgHom K R) = (· ^ q) := rfl
+
+/-- If R is a perfect ring and an algebra over a finite field K, the Frobenius K-algebra
+  endomorphism of R is an automorphism. -/
+@[simps!] noncomputable def frobeniusAlgEquiv (p : ℕ) [ExpChar R p] [PerfectRing R p] : R ≃ₐ[K] R :=
+  .ofBijective (frobeniusAlgHom K R) <| show Function.Bijective (· ^ q) by
```

Comment:
Maybe it is unavoidable, but `show` is a little code smell.

Reply:
This is easy to eliminate ... my negligence